caddy: Allow adding external plugins

jpds opened the pull request:
## Description of changes

Solves: #14671

Caddy has support for plugins that are added at compile time. This exposes an argument `externalPlugins` that will build caddy with the specified plugins.

Example:
```
caddy-custom = pkgs.caddy.override {
  externalPlugins = [
    {name = "coraza"; repo = "github.com/corazawaf/coraza-caddy/v2"; version = "v2.0.0-rc.3";}
    {name = "cache-handler"; repo = "github.com/caddyserver/cache-handler"; version = "v0.9.0";}
    {name = "ratelimit"; repo = "github.com/mholt/caddy-ratelimit"; version = "2dc0d586f0b87e983757c403bc0929ddeb84a537";}
  ];
  vendorHash = "<SRI hash>";
};
```

Adapted from Brendan Taylor's CoreDNS change in commit 95e66809debf42dbe1e4935fd31c8c275914d2eb.

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [X] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [X] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [X] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

emilylange commented:
While I do see that something has to happen to solve this issue, I don't feel like exposing `vendorHash` to end-users is the way to go :(

jpds committed:
caddy: Allow adding external plugins

Solves: #14671

Caddy has support for plugins that are added at compile time. This exposes an
argument `externalPlugins` that will build caddy with the specified plugins.

Example:
```
caddy-custom = pkgs.caddy.override {
  externalPlugins = [
    {name = "coraza"; repo = "github.com/corazawaf/coraza-caddy/v2"; version = "v2.0.0-rc.3";}
    {name = "cache-handler"; repo = "github.com/caddyserver/cache-handler"; version = "v0.9.0";}
    {name = "ratelimit"; repo = "github.com/mholt/caddy-ratelimit"; version = "2dc0d586f0b87e983757c403bc0929ddeb84a537";}
  ];
  vendorHash = "<SRI hash>";
};
```

Adapted from Brendan Taylor's CoreDNS change in commit 95e66809debf42dbe1e4935fd31c8c275914d2eb.

jpds commented:
Ultimately, a hash needs to be presented to the user somewhere.

For those who just use Caddy, life continues as normal; those of us who require plugins will of course have to deal with the dance to getting the new hash everytime - but that's an OK trade-off as it's all managed by nix in the end (which is significantly better than the alternative of random binaries made by `xcaddy`).

lorenzleutgeb commented:
Is it possible to resolve this by creating fixed-output derivations for the external plugins, and "injecting" them into the `buildGoModule`-based derivation?

```nix
mercure = buildGoModule rec {
  pname = "mercure";
  version = "0.15.5";
  src = fetchFromGitHub {
    owner = "dunglas";
    repo = "mercure";
    rev = "v${version}";
    hash = "sha256-DyKNKhxjnOfxYcp3w1nB6kxs9c4ZaHL0AN0Eb5vc6mA=";
  };
  subPackages = [ "." ];
  phases = [ "unpackPhase" "patchPhase" "configurePhase" "buildPhase" ];
  buildPhase = "cp -R . $out";
  vendorHash = "sha256-ZubgU8X7zPgzdFo1/aM8pNCCY29+WjVH+pVSll1lUlE=";
}
```

@augustebaum @imincik @kubaneko and me tried to get this to work, but unfortunately we do not understand enough about `buildGoModule` works. We think that it would be reasonable to add output from derivations as additional inputs for building a Go module. Just adding the dependencies to `vendor/` would again result in changes to `vendorHash`, but maybe there's a trick involving `proxyVendor = true`? Maybe it even requires changes to `buildGoModule`?

If that's not possible, then your solution is a good compromise between exposing hashes to the user and complexity of a clean implementation.

elsbrock commented:
Using

```
  services.caddy = {
    enable = true;
    package = (pkgs.callPackage ./custom-caddy.nix {
      externalPlugins = [
        {name = "caddy-security"; repo = "github.com/greenpau/caddy-security"; version = "v1.1.20"; }
      ];
      vendorHash = "sha256-YNcQtjPGQ0XMSog+sWlH4lG/QdbdI0Lyh/fUGqQUFaY=";
    });
  };
```

Getting:

```
@nix { "action": "setPhase", "phase": "unpackPhase" }
unpacking sources
unpacking source archive /nix/store/6qhv0b3ripxgjyw8jhibqmn1i3r91bzs-source
source root is source
@nix { "action": "setPhase", "phase": "patchPhase" }
patching sources
@nix { "action": "setPhase", "phase": "updateAutotoolsGnuConfigScriptsPhase" }
updateAutotoolsGnuConfigScriptsPhase
@nix { "action": "setPhase", "phase": "configurePhase" }
configuring
@nix { "action": "setPhase", "phase": "buildPhase" }
building
Building subPackage ./cmd/caddy
go: finding module for package github.com/greenpau/caddy-security
cmd/caddy/main.go:36:1: cannot query module due to -mod=vendor
```

jpds commented:
> Is it possible to resolve this by creating fixed-output derivations for the external plugins, and "injecting" them into the buildGoModule-based derivation?

I feel that this is what the `caddy-go-modules` derivation already does(?).

jan4843 commented:
Is `name` in each `externalPlugins` used? It seems it could be dropped.

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-ready-for-review/1016/90

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-ready-for-review/3032/3315

h7x4 commented on pkgs/servers/caddy/default.nix:
```diff
@@ -35,7 +42,34 @@ buildGoModule {
+    for module in ${builtins.toString (attrsToModules externalPlugins)}; do
+      sed -i "/standard/a _ \"$module\"" ./cmd/caddy/main.go
+    done
+    for plugin in ${builtins.toString (attrsToSources externalPlugins)}; do
```
```suggestion
    for plugin in ${toString (attrsToSources externalPlugins)}; do
```

h7x4 commented on pkgs/servers/caddy/default.nix:
```diff
@@ -35,7 +42,34 @@ buildGoModule {
+  nativeBuildInputs = [ gnused installShellFiles ];
+
+  modBuildPhase = ''
+    for module in ${builtins.toString (attrsToModules externalPlugins)}; do
+      sed -i "/standard/a _ \"$module\"" ./cmd/caddy/main.go
+    done
```
We could probably do without `sed` here and further below. Something like this maybe?

```suggestion
    substituteInPlace ./cmd/caddy/main.go \
      --replace "  // plug in Caddy modules here" "${
        lib.concatMapStringsSep "\\n"
          (module: "  _ ${module}")
          (attrsToModules externalPlugins)
      }"
```

h7x4 commented:
@lorenzleutgeb it does seem doable, but I think it would require changes to `buildGoModule`. The only FOD in the go builder seems to be [`goModules`](https://github.com/NixOS/nixpkgs/blob/d49a222c10ff9dd531b4e85f28cb9ae42d6daad5/pkgs/build-support/go/module.nix#L58), and it seems like the only time it accesses the internet is here:

https://github.com/NixOS/nixpkgs/blob/d49a222c10ff9dd531b4e85f28cb9ae42d6daad5/pkgs/build-support/go/module.nix#L111

We could make a function that downloads sources individually or in a batch (maybe something like `fetchGoDeps`?), which exposes the normal override functions for this derivation to allow for combining it with other derivation. Then, we use this in a new derivation that runs `go mod vendor`, however this time without the need for a `vendorHash`. As long as `go mod vendor` doesn't need to access the internet at least?

SuperSandro2000 commented on pkgs/servers/caddy/default.nix:
```diff
@@ -35,7 +42,34 @@ buildGoModule {
   ];
 
-  nativeBuildInputs = [ installShellFiles ];
+  nativeBuildInputs = [ gnused installShellFiles ];
```
```suggestion
  nativeBuildInputs = [ installShellFiles ];
```
gnused is part of stdenv and does not need to be added

h7x4 replied:
Wouldn't gnused be needed to ensure `-i` works on darwin and bsd? There are some flag incompatibilities: https://stackoverflow.com/questions/5694228/sed-in-place-flag-that-works-both-on-mac-bsd-and-linux

Mayeu replied:
Yes, `sed -i` without an explicit extension (`-i.bak`) won't work on macOS. So we either have to keep `gnused`, or change the sed command on line 49.

SuperSandro2000 commented on nixos/doc/manual/release-notes/rl-2311.section.md:
```diff
@@ -501,6 +501,26 @@ The module update takes care of the new config syntax and the data itself (user
+    enable = true;
+    package = pkgs.caddy.override {
+      externalPlugins = [
+        {name = "coraza"; repo = "github.com/corazawaf/coraza-caddy/v2"; version = "v2.0.0-rc.3";}
+        {name = "cache-handler"; repo = "github.com/caddyserver/cache-handler"; version = "v0.9.0";}
+        {name = "ratelimit"; repo = "github.com/mholt/caddy-ratelimit"; version = "2dc0d586f0b87e983757c403bc0929ddeb84a537";}
```
```suggestion
        { name = "coraza"; repo = "github.com/corazawaf/coraza-caddy/v2"; version = "v2.0.0-rc.3"; }
        { name = "cache-handler"; repo = "github.com/caddyserver/cache-handler"; version = "v0.9.0"; }
        { name = "ratelimit"; repo = "github.com/mholt/caddy-ratelimit"; version = "2dc0d586f0b87e983757c403bc0929ddeb84a537"; }
```

SuperSandro2000 commented on nixos/doc/manual/release-notes/rl-2311.section.md:
````diff
@@ -501,6 +501,26 @@ The module update takes care of the new config syntax and the data itself (user
+  };
+  ```
+
+  To get the necessary SRI hash, set `vendorHash = "";`. The build will fail and produce the correct `vendorHash` in the error message.
````
```suggestion
  To get the necessary SRI hash for the vendor directory, first set `vendorHash = "";`. Than build the next build will fail and log the correct `vendorHash` in the error message.
```

SuperSandro2000 commented on nixos/doc/manual/release-notes/rl-2311.section.md:
```diff
@@ -501,6 +501,26 @@ The module update takes care of the new config syntax and the data itself (user
+
+  To get the necessary SRI hash, set `vendorHash = "";`. The build will fail and produce the correct `vendorHash` in the error message.
+
+  If you use this feature, updates to Caddy may require updating `vendorHash` by following these steps again.
```
```suggestion
  If you use this feature, updates to Caddy likely require updating `vendorHash` by following these steps again.
```

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-already-reviewed/2617/1473

emilylange commented:
hm... alright, let's give this another go (as one of the caddy maintainers in nixpkgs):

I see and acknowledge, that there is need for this. There has been for years.
And I also agree that maintaining out-of-tree variants of this is bad UX.

Upstream recently reignited the discussion regarding Go plugin support in https://github.com/caddyserver/caddy/issues/5183#issuecomment-1872557403, which turned out better than the year(s) before.
But it's still unclear, a big undertaking and low-prio, so we don't want to wait for it.

I do, however, not like, that this replaces current caddy package.
And I don't like how it exposes raw FOD hashes, but this is just how it is -- at least for now.

> which is significantly better than the alternative of random binaries made by `xcaddy`

`xcaddy` has multiple modes and does not have to build the binary. It can also bootstrap the file structure for us using `XCADDY_SKIP_BUILD=1`. This is also deterministic and reproducible.

And example for this, even if it needs some cleanup, e.g. removing `with pkgs;`, can be found here: https://github.com/emilylange/nixos-config/blob/f49eaf2f1deca07eb0dc76d0ae426c9651b1d446/packages/caddy/default.nix

---

How about the following:

We keep the current `caddy` drv as is, and add a `with-external-plugins` in `passthru`.

Users who need external plugins can use that via `caddy.with-external-plugins {}` and don't have to do the whole `.override` dance.

This also comes with the added benefit, that the `vendorHash` can be left empty by default to use `lib.fakeHash` instead of the vanilla caddy-without-plugins one.
Meaning there is less confusion if one forgets to override that.

 I think I would prefer to use the `XCADDY_SKIP_BUILD` approach I shared earlier, but I don't have a very strong opinion on this.

Denperidge commented:
Here's a way of implementing it from the fork without having to create a file locally!

```nix 
services.caddy = {
  enable = true;   package = (pkgs.callPackage "${builtins.fetchurl https://raw.githubusercontent.com/jpds/nixpkgs/a33b02fa9d664f31dadc8a874eb1a5dbaa9f4ecf/pkgs/servers/caddy/default.nix}" {
    externalPlugins = [
      { name = "caddy-dns/porkbun"; repo = "github.com/caddy-dns/porkbun"; version = "4267f6797bf6543d7b20cdc8578a31764face4cf"; }
    ];
    vendorHash = "";  # Add this, as explained in https://github.com/NixOS/nixpkgs/pull/259275#issuecomment-1763478985 
  });
  globalConfig = ''
    ...
  '';
}
  ```

dbaynard commented:
> I think I would prefer to use the `XCADDY_SKIP_BUILD` approach I shared earlier, but I don't have a very strong opinion on this.

This does look quite nice — `xcaddy` is the upstream-supported way to do this so there's less maintenance burden with this approach?

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-already-reviewed/2617/1603

kanashimia commented on pkgs/servers/caddy/default.nix:
```diff
@@ -35,7 +42,34 @@ buildGoModule {
+
+  modInstallPhase = ''
+    mv -t vendor go.mod go.sum
+    cp -r --reflink=auto vendor "$out"
```
`--reflink=auto` is superfluous, it is the default since coreutils 9

kanashimia commented:
It seems that both this PR and approach with xcaddy use `go mod vendor`, with means that you can't use them to build something like `frankenphp`, as it requires C libs that are removed upon vendoring, `buildGoModule` already has `proxyVendor = true;` to handle such cases, as I see you would require to implement that logic here too, which isn't very nice.

Nanotwerp commented:
> It seems that both this PR and approach with xcaddy use `go mod vendor`, with means that you can't use them to build something like `frankenphp`, as it requires C libs that are removed upon vendoring, `buildGoModule` already has `proxyVendor = true;` to handle such cases, as I see you would require to implement that logic here too, which isn't very nice.

`frankenphp` is already packaged, though. The PR is more than enough for most `Caddy` plugins; `frankenphp` is a bit more than a normal plugin, and acts as its own server, even though it's technically built atop `Caddy`.

kanashimia commented:
> `frankenphp` is already packaged, though. The PR is more than enough for most `Caddy` plugins; `frankenphp` is a bit more than a normal plugin, and acts as its own server, even though it's technically built atop `Caddy`.

Yes, but like that it would require you to basically have two caddys running, and you won't be able to add extensions to one of them.

In any case my point was that plugins with C deps need to be looked more into, as there is quite a bit of them.
After looking into it I couldn't find at a glance another plugin that would break, so it isn't really important.

It actually seems that `go mod vendor` works in most of the cases, I've tried to build https://github.com/mliezun/caddy-snake and it works. The problem occurs only when there is C code in the subdirectories, as per https://github.com/golang/go/issues/26366, when you move C code to the root dir it works, frankenphp could just be "fixed" then. C-Thread-Pool code that frankenphp vendors can be found in a separate repo too, I was able to build frankenphp with it.

flokli commented:
Even if we might need to apply some more fixes for plugins with C dependencies, it wouldn't change any of the FOD digests of the non-C plugins, right?

So we could land this without solving that particular issue, and worst case we'd need to update FOD digests for plugins with C as we update the packaging expressions regarding that. 

Can this be rebased?

Nanotwerp commented:
I created a [PR](github.com/jpds/nixpkgs/pull/1) on @jpds's branch for this PR that should help with rebasing. The PR is based on a branch I made that's rebased. My PR depends on #316076, which will update `caddy` . After #316076 is merged, they can reset their branch to the commit of the update and then merge my PR to their branch. That should hopefully rebase this PR.

phanirithvij commented:
https://github.com/NixOS/nixpkgs/pull/358586 was merged. fyi
